### DISCUSSION TITLE
GC in the context of game development with a possible solution via Fiber local variables (mRuby specific).

### DISCUSSION THREAD

**Discussion by amirrajan** [ioquatix/ioquatix#36] [Q&A]
There are a few memory management mechanisms that games can benefit from and I'm trying to get ideas about how these different memory management needs can be accomplished in mRuby.

# Mark and Sweep
This is obviously handled already and is a boon :-)

# Never GC
Long-running "constant" state is another facet of games. The primary scenario here would be a large, tile-based terrain/stage definition.

When a game starts, a large dataset gets loaded from some persisted storage. After this data is loaded, it is never changed, and never deleted from memory. The amount of data that gets loaded can be very large. An `Array` with 2.5 million `Hash` objects is totally within the realm of "reasonable".

The problem with this is the array - being alive along with all its children - significantly hurts full gc. 

Assuming a simulation loop running at 60hz (60 fps). You have a 16ms time slice to process game logic. Ideally, you don't want incremental gc to occur within that 16ms time slot. Doing a full GC after the simulation loop, has worked well. But, with a large dataset being alive, a full gc takes too long.

# Always GCed
There is also a partition of memory that is categorized as "temp state". Objects allocated within this arena are disposed of in their entirety at the end of the simulation tick. This arena has a fixed size which makes it extremely fast (move the pointer to the heap back to the beginning after simulation executes).

# Local Variables in Fibers
With respect to local variables in fibers, are they checked for GC? I understand that the Fiber's implicit block isn't a scope gate, but if the variable is truly local, does it have to be checked during mark and sweep? Could we implicitly assume that local variables within the Fiber's context are alive if the Fiber is alive?

If this seems like a reasonable optimization for Fibers, this would solve the "Never GC" scenario possibly (granted that it's kind of a hack):

```ruby
@fiber = Fiber.new do
  big_object = #Array with 2M hashes
  while true
    Fiber.yield big_object
  end
end

# elsewhere if bo is needed
bo = @fiber.resume
```

As for the "Always GCed" scenario, I don't have any ideas.

**Comment by ioquatix** [comment]
As far as I know, nothing like this exists, and GC can be a major problem for concurrency/parallelism in a similar way that it's a problem for game implementations. It seems unfortunate as these problems are trivial in a language with manual memory management (ignoring all the other issues). Maybe we can introduce a specific data type into Ruby which avoids some of the overheads.

> With respect to local variables in fibers, are they checked for GC?

To answer your question directly, the answer is yes, the same as any other stack.

> Could we implicitly assume that local variables within the Fiber's context are alive if the Fiber is alive?

In theory, it should be only necessary to scan the parts of the stack (and heap) that have changed. e.g. we worked on this: https://github.com/ruby/ruby/pull/8365

Realistically though, I think a more advanced GC is required, ideally one which doesn't stop the world or minimises stopping the world.

**Comment by amirrajan** [comment]
> It seems unfortunate as these problems are trivial in a language with manual memory management

For the specific case of "GC everything in this reserved arena", it's definitely simple and has some solid benefits.

> Maybe we can introduce a specific data type into Ruby which avoids some of the overheads.

If there was Array/Hash construct that indiscriminately GCed all its content (with a full sweep ignoring this construct's children), that would be super helpful for the object pooling scenario. It's a very niche usecase though (since we have a very well defined point in the game lifecycle where the Ruby VM is essentially "shut down"/not used).

**Comment by Asmod4n** [comment]
Hm, you can plug your own memory allocation functions in mruby, practically ignoring any memory management at all and just allocate a large enough memory region up front. And you can also disable the GC permanently in mruby.

You could also go so far and implement manual memory management for mruby, the tools should all be there.
